src/mlpack/methods/ann/layer/layer.hpp

```diff
@@ -12,73 +12,312 @@
-#include "highway.hpp"
-#include "instance_norm.hpp"
-#include "join.hpp"
-#include "layer_norm.hpp"
-#include "layer_types.hpp"
-#include "leaky_relu.hpp"
-#include "linear.hpp"
-#include "linear_no_bias.hpp"
-#include "linear3d.hpp"
-#include "log_softmax.hpp"
-#include "lookup.hpp"
-#include "lp_pooling.hpp"
-#include "lstm.hpp"
-#include "max_pooling.hpp"
-#include "mean_pooling.hpp"
-#include "minibatch_discrimination.hpp"
-#include "multihead_attention.hpp"
-#include "multiply_constant.hpp"
-#include "multiply_merge.hpp"
-#include "nearest_interpolation.hpp"
-#include "noisylinear.hpp"
-#include "padding.hpp"
-#include "parametric_relu.hpp"
-#include "pixel_shuffle.hpp"
-#include "positional_encoding.hpp"
-#include "recurrent_attention.hpp"
-#include "recurrent.hpp"
-#include "reinforce_normal.hpp"
-#include "relu6.hpp"
-#include "reparametrization.hpp"
-#include "select.hpp"
-#include "sequential.hpp"
-#include "softshrink.hpp"
-#include "softmax.hpp"
-#include "softmin.hpp"
-#include "spatial_dropout.hpp"
-#include "subview.hpp"
-#include "transposed_convolution.hpp"
-#include "virtual_batch_norm.hpp"
-#include "vr_class_reward.hpp"
-#include "weight_norm.hpp"
+namespace mlpack {
+namespace ann {
+
+/**
+ * A layer is an abstract class implementing common neural networks operations,
+ * such as convolution, batch norm, etc. These operations require managing
+ * weights, losses, updates, and inter-layer connectivity.
+ *
+ * Users will just instantiate a layer by inherited from the abstract class and
+ * implement the layer specific methods. It is recommend that descendants of
+ * Layer implement the following methods:
+ *
+ *  - Constructor: Defines custom layer attributes, and creates layer state
+ *    variables.
+ *
+ *  - Forward(input, output): Performs the forward logic of applying the layer
+ *    to the input object and storing the result in the output object.
+ *
+ *  - Backward(input, gy, g): Performs a backpropagation step through the layer,
+ *    with respect to the given input.
+ *
+ *  - Gradient(input, error, gradient): Computing the gradient of the layer with
+ *    respect to its own input.
+ *
+ * The memory for the layer's parameters (weights and biases) is not allocated
+ * by the layer itself, instead it is allocated by the network that the layer
+ * belongs to, and passed to the layer when it needs to use it.
+ *
+ * See the linear layer implementation for a basic example. It's a layer with
+ * two variables, w and b, that returns y = w * x + b. It shows how to implement
+ * Forward(), Backward() and Gradient().  The weights of the layers are tracked
+ * in layer.Parameters().
+ *
+ * @tparam MatType Matrix representation to accept as input and use for
+ *    computation.
+ */
+template<typename MatType = arma::mat>
+class Layer
+{
+ public:
+  //! Default constructor.
+  Layer() : validOutputDimensions(false), training(false)
+  { /* Nothing to do here */ }
+
+  //! Default deconstructor.
+  virtual ~Layer() { }
```

Comment:
```suggestion
  virtual ~Layer() { /* Nothing to do here */ }
```

To be consistent with the other methods from this class.